### ISSUE TITLE
TwigRenderer underlying twig environment getter

### ISSUE THREAD

**Issue by n3vrax** [zendframework/zend-expressive-twigrenderer#14]
Hi,

I was put in the situation where I need to access the twig environment of the TwigRenderer for further configuration, like registering a fallback function. As far as I can see, there is no getter for it, so I find myself in the situation where I need to rewrite the entire TwigRendererFactory for it to work.
Any ideas, or maybe you could include this feature in next releases?

I would like to be able to do something like

``` php
if($template instanceof TwigRenderer) {
  $env = $template->getTwigEnvironment();
  $env->...
}
```

Thanks!

**Comment by geerteltink** [comment]
I ran in the same situation and there are a few solutions for this:
1. Rewrite the entire TwigRenderFactory as you said
2. Write your own extension
3. Write a PR and add `getTwigEnvironment`
4. Write a PR to split the `TwigEnvironment` from the TwigRendererFactory

Option 3 or 4 would be great but I don't know which way would be best. I'm think I prefer option 4 over option 3 so the Twig_Environment can be reused anyway we want.

But at the end I choose option 2:

``` php
class FlashMessengerExtension extends \Twig_Extension implements \Twig_Extension_GlobalsInterface
{
    private $flashMessenger;

    public function setFlashMessenger(FlashMessenger $flashMessenger)
    {
        $this->flashMessenger = $flashMessenger;
    }

    public function getGlobals()
    {
        return [
            'app' => [
                'flash' => $this->flashMessenger
            ]
        ];
    }

    public function getName()
    {
        return 'flash-messenger';
    }
}
```

In this case the FlashMessenger is initialized in a Middleware and injected into the extension. By the time the app gets through the middleware stack and reaches the ActionMiddleware, the extension is ready and the TwigRendererFactory uses the extension. Obviously the extension was already added in the config to `['twig']['extensions']`.

EDIT:

I also needed to access the twig environment to extract all translation strings with a cli command. The easiest way to do that was with Reflection:

``` php
$reflectionClass    = new ReflectionClass(TwigRenderer::class);
$reflectionProperty = $reflectionClass->getProperty('template');
$reflectionProperty->setAccessible(true);

$environment = $reflectionProperty->getValue($this->templateRenderer);
```

**Comment by n3vrax** [comment]
Thanks for you fast reply!
I'll be going with 1 for now, because I don't see how 2 will help in my case

Basically, I needed to make available the zend-view helpers in twig template without rewriting extensions for them
So in the TwigRendererFactory, I've added this to __invoke, and it seems to work, I can access zend form view helpers as needed

``` php
$viewHelperManager = $app->serviceManager->get('ViewHelperManager');
$renderer = new \Zend\View\Renderer\PhpRenderer();
$renderer->setHelperPluginManager($viewHelperManager);

$environment->registerUndefinedFunctionCallback(
    function ($name) use ($viewHelperManager, $renderer) {
        if (!$viewHelperManager->has($name)) {
            return false;
        }

        $callable = [$renderer->plugin($name), '__invoke'];
        $options  = ['is_safe' => ['html']];
        return new \Twig_SimpleFunction(null, $callable, $options);
    }
);
```

I'll write a PR when I get to it
Thanks!

**Comment by n3vrax** [comment]
Sorry, haven't read your edit
Yeah, that looks like a better solution than rewriting factory, I'll give it a try
Thanks

**Comment by weierophinney** [comment]
Another possibility is that we add a factory for the `TwigEnvironment`. If we were to do so, users could add delegator factories to that service in order to further customize it, and our own functionality for injecting extensions could itself be done as a delegator factory. @xtreamwayz — thoughts?

**Comment by geerteltink** [comment]
@weierophinney That is my 4th option and also the one I prefer. It would definitely make life easier to extend Twig on the fly to reuse its configuration where needed.

I don't think you need delegator factories though. As long as you can get the TwigEnvironment from the container, you can extend it the way you want within its own api.

**Comment by weierophinney** [comment]
This is resolved with #15.
